pkg/nvidia/nvfanctrl/main.go

```diff
@@ -0,0 +1,366 @@
+		points: [][]int{
+			{0,66},
+			{10,66},
+			{11,215},
+			{30,215},
+			{60,255},
+			{105,255},
+		},
+    }
+	return modeCool
+}
+
+// Check if the temperature file exist for a thermal zone
+func checkThermalZone(tzone int) bool {
+	if _, err := os.Stat(getThermalDevice(tzone)); err == nil {
+		return true
+	}
+	return false
+}
+
+// Return the thermal zone temperature file
+func getThermalDevice(tzone int) string {
+	return fmt.Sprintf(thermalDevice, tzone)
+}
+
+// Identify device model and return PWM profile accordingly
+// Reference: https://docs.nvidia.com/jetson/archives/r35.4.1/DeveloperGuide/text/HR/JetsonEepromLayout.html
+func getDeviceProfile() (deviceProfiles, error) {
+	asset, err := ioutil.ReadFile(assetTagSysFs)
+	if err != nil {
+		return nil, err
+	}
+	assetTag := string(asset)
+
+	// Extract model from Asset TAG
+	reAT := regexp.MustCompile("699-[0-9]([0-9]{4})-([0-9]{4})-([0-9]{3})( .*)")
+	if !reAT.MatchString(assetTag) {
+		return nil, fmt.Errorf("Asset TAG doesn't match any known model")
+	}
+	str := strings.TrimSpace(reAT.ReplaceAllString(assetTag, "$1"))
+
+	// Decode model
+	devIDs := map[string]deviceProfiles{
+		"3767": profileOrin{},
+		"3701": profileOrin{},
+		"3768": profileOrin{},
+		"3737": profileOrin{},
+		"3668": profileXavier{},
+		"2888": profileXavier{},
+		"3509": profileXavier{},
+		"2822": profileXavier{},
+	}
+	dev, ok := devIDs[str]
+	if !ok {
+		return nil, fmt.Errorf("Unknown Jetson model")
+	}
+
+	return dev, nil
+}
+
+// Search for the PWM Fan device
+func getFANDevice() (string, error) {
+	entries, err := os.ReadDir(fanSysDir)
+	if err != nil {
+		return "", err
+	}
+	// We just need the first entry (most probably the only one)
+	if len(entries) >= 1 {
+		dev := fanSysDir + entries[0].Name() + "/pwm1"
+		return dev, nil
+	}
+	return "", nil
+}
+
+// Read and convert (to Celsius) the temperature from a thermal device
+func (pwm fanCTRL) readTemp() (int, error) {
+	data, err := os.ReadFile(pwm.devThermal)
+	if err != nil {
+		return -1, err
+	}
+
+	str := string(data[:len(data)-1])
+	temp, errconv := strconv.Atoi(str)
+	if errconv != nil {
+		return -1, errconv
+	}
+
+	return temp/1000, nil
+}
+
+// Set PWM value of a FAN device
+func (pwm fanCTRL) setPWM(pwmValue int) error {
+	errF := os.WriteFile(pwm.devPWM, []byte(strconv.Itoa(pwmValue)), 0644)
+	if errF != nil {
+		return fmt.Errorf("Fail to change FAN speed")
+	}
+	return nil
+}
+
+// Perform the controlling iteration
+func (pwm fanCTRL) controlPWM() error {
+	temp, errT := pwm.readTemp()
+	if errT != nil {
+		return fmt.Errorf("Fail to read temperature!")
+	}
+
+	pwmValue := maxPWM // Set to maximum
+	for _, v := range pwm.cmode.points {
+		if temp < v[0] {
+			pwmValue = v[1]
+			break
+		}
+	}
+
+	return pwm.setPWM(pwmValue)
+}
+
+// Controlling loop
+func (pwm fanCTRL) run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-time.Tick(time.Duration(pwm.checkInterval) * time.Second):
```

Comment:
As far as I understand, it creates a new channel and a goroutine every time. And it's a resource leak. As far as I understand that's why we usually create a ticker once and then wait the ticks in a loop, like that:
```go
func (pwm fanCTRL) run(ctx context.Context) error {
	ticker := time.NewTicker(time.Duration(pwm.checkInterval) * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return nil

		case <-ticker.C:
			err := pwm.controlPWM()
			if err != nil {
				fmt.Println("Error:", err)
			}
		}
	}
}
```
 @christoph-zededa, am I right?

Reply:
I am not sure if it is really a resource leak, I think channels get garbage-collected even if not closed.

Interesting: 
> As of Go 1.23, the garbage collector can recover unreferenced tickers, even if they haven't been stopped. The Stop method is no longer necessary to help the garbage collector. There is no longer any reason to prefer NewTicker when Tick will do. 

https://pkg.go.dev/time#Tick

@rene is ahead of us ;-)

Reply:
@christoph-zededa can we run some magic go tool to verify the leak here? If yes, do we have it somewhere in our pipeline?

Reply:
https://pkg.go.dev/net/http/pprof can be used for things like this.

Or the profile can created by running a test with `-memprofile=output.profile` like f.e. for pillar: https://github.com/lf-edge/eve/blob/master/pkg/pillar/build-scripts/memprof_test.sh#L14

Reply:
you are right @OhmSpectator , we had this leak in the past, I will fix it (since we are not on GO 1.23)...